File: clients/python/src/dbctl/options.py
DEFAULT_HOST_ADDRESS = "localhost"
DEFAULT_HOST_PORT = 1988

# The instance details default to empty: the server knows the defaults of each
# database type and fills in the missing ones. Sending the postgres defaults for
# every type would make dbctl try to log into redis as "postgres".
DEFAULT_INSTANCE_PORT = 0
DEFAULT_INSTANCE_USER = ""
DEFAULT_INSTANCE_PASS = ""
DEFAULT_INSTANCE_DB_NAME = ""


class Config:
    migrations: str
    migrations_file_regex: str
    fixtures: str

    with_default_migrations: bool

    instance_port: int
    instance_user: str
    instance_pass: str
    instance_db_name: str

    host_address: str
    host_port: int

    def __init__(self, **kwargs):
        # empty by default: a path is only sent when the caller asks for it, so a
        # missing directory is reported instead of silently producing an empty
        # database.
        self.migrations = kwargs.get("migrations", "")
        self.migrations_file_regex = kwargs.get("migrations_file_regex", "")
        self.fixtures = kwargs.get("fixtures", "")

        self.with_default_migrations = kwargs.get("with_default_migrations", False)

        self.instance_port = kwargs.get("instance_port", DEFAULT_INSTANCE_PORT)
        self.instance_user = kwargs.get("instance_user", DEFAULT_INSTANCE_USER)
        self.instance_pass = kwargs.get("instance_pass", DEFAULT_INSTANCE_PASS)
        self.instance_db_name = kwargs.get("instance_db_name", DEFAULT_INSTANCE_DB_NAME)

        self.host_address = kwargs.get("host_address", DEFAULT_HOST_ADDRESS)
        self.host_port = kwargs.get("host_port", DEFAULT_HOST_PORT)

    def use_default_migrations(self):
        """Reuse the migrations the instance was started with."""
        self.with_default_migrations = True
        return self

    def with_migrations(self, migrations: str, file_regex: str = ""):
        self.migrations = migrations
        self.migrations_file_regex = file_regex
        return self

    def with_fixtures(self, fixtures: str):
        self.fixtures = fixtures
        return self

    def with_instance(self, user: str, password: str, dbname: str, port: int):
        self.instance_user = user
        self.instance_pass = password
        self.instance_db_name = dbname
        self.instance_port = port
        return self

    def with_host(self, address: str, port: int):
        self.host_address = address
        self.host_port = port
        return self

    def get_host_url(self):
        return f"http://{self.host_address}:{self.host_port}"

    # kept for backwards compatibility with the singular names
    with_migration = with_migrations
    with_fixture = with_fixtures

File: clients/dbctlgo/options.go
package dbctlgo

import (
	"fmt"
	"net"
	"os"
	"strconv"
)

// config is the client configuration.
type config struct {
	migrations          string
	migrationsFileRegex string

	fixtures string

	// whether or not to use default migrations/fixtures loaded when dbctl started
	withDefaultMigrations bool

	// postgres instance information
	instancePort   uint32
	instanceUser   string
	instancePass   string
	instanceDBName string

	// host and port of the host, where the dbctl testing server is running
	hostAddress string
	hostPort    uint32
}

// Environment variables read when no option overrides them, so that a project can
// point its tests at its own dbctl server without touching the test code.
const (
	// EnvHost is the address of the dbctl api server
	EnvHost = "DBCTL_HOST"
	// EnvPort is the port of the dbctl api server
	EnvPort = "DBCTL_PORT"
)

// The instance details are left empty on purpose: the server knows the defaults of
// each database type and fills in the missing ones. Sending the postgres defaults
// for every type would make dbctl try to log into redis as "postgres".
var defaultConfig = &config{
	hostAddress: "localhost",
	hostPort:    1988,
}

// fromEnv applies the environment to a config, before any explicit option.
func (c *config) fromEnv() error {
	if host := os.Getenv(EnvHost); host != "" {
		c.hostAddress = host
	}

	if port := os.Getenv(EnvPort); port != "" {
		p, err := strconv.ParseUint(port, 10, 32)
		if err != nil {
			return fmt.Errorf("%s is not a valid port number: %q", EnvPort, port)
		}
		c.hostPort = uint32(p)
	}

	return nil
}

// Option is a function that configures the client.
type Option func(*config) error

// WithMigrations configures the client to use the given migrations.
func WithMigrations(migrations string) Option {
	return func(cfg *config) error {
		cfg.migrations = migrations
		return nil
	}
}

// WithDefaultMigrations configures the client to use the default migrations.
func WithDefaultMigrations() Option {
	return func(cfg *config) error {
		cfg.withDefaultMigrations = true
		return nil
	}
}

// WithFixtures configures the client to use the given fixtures.
func WithFixtures(fixtures string) Option {
	return func(cfg *config) error {
		cfg.fixtures = fixtures
		return nil
	}
}

// WithInstance configures the client to use the given postgres instance.
// The address is accepted for backwards compatibility and ignored: the dbctl
// server always reaches the instances running next to it.
func WithInstance(user, pass, address, dbname string, port uint32) Option { //nolint:revive
	return func(cfg *config) error {
		cfg.instanceUser = user
		cfg.instancePass = pass
		cfg.instanceDBName = dbname
		cfg.instancePort = port
		return nil
	}
}

// WithHost configures the client to use the given host.
func WithHost(address string, port uint32) Option {
	return func(cfg *config) error {
		cfg.hostAddress = address
		cfg.hostPort = port
		return nil
	}
}

// WithMigrationsFileRegex configures the client to use the given regex to match what migrations file to apply.
func WithMigrationsFileRegex(regex string) Option {
	return func(cfg *config) error {
		cfg.migrationsFileRegex = regex
		return nil
	}
}

// getHostURL returns the host url.
func (c *config) getHostURL() string {
	return "http://" + net.JoinHostPort(c.hostAddress, fmt.Sprintf("%d", c.hostPort))
}
